### ISSUE TITLE
Backend API not working as expected

### ISSUE THREAD

**Issue by aswad546** [wspr-ncsu/visiblev8-crawler#1]
Hello,

I followed the instructions in your repository as mentioned and installed the required dependencies. But when I attempt to crawl a webpage I get a server error when it tries to send data to your backend at route /urlsubmit. Is it possible there are some dependencies missing from your requirements.txt file?

Best,
Aswad

**Comment by sohomdatta1** [comment]
@aswad546 Could you post the error you did get ?

**Comment by sohomdatta1** [comment]
Also what configurations did you use while setting it up?

**Comment by aswad546** [comment]
Apologies for not sharing before I am using ubuntu 22.0.4 (jammy) with a sufficient amount of RAM (256),
Here is the error trace:
It is caused due to an internal server error in your backend API. 

```
`Traceback (most recent call last):
  File "/home/vagrant/.local/lib/python3.10/site-packages/requests/models.py", line 971, in json
    return complexjson.loads(self.text, **kwargs)
  File "/usr/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/vagrant/visiblev8-crawler/./scripts/vv8-cli.py", line 50, in <module>
    main()
  File "/home/vagrant/visiblev8-crawler/./scripts/vv8-cli.py", line 41, in main
    crawl.crawler(opts, unkown_args)
  File "/home/vagrant/visiblev8-crawler/scripts/crawl.py", line 132, in crawler
    crawler_inst.crawl([ args.url ])
  File "/home/vagrant/visiblev8-crawler/scripts/crawl.py", line 106, in crawl
    submission_id = r.json()['submission_id']
  File "/home/vagrant/.local/lib/python3.10/site-packages/requests/models.py", line 975, in json
    raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)`
```

**Comment by sohomdatta1** [comment]
Hmm, that looks like a generic backend failure message,

What options did choose when you used the setup script? Did you choose `local`, `connect`, or `remote` and how many instances did you choose?

Also, if you run `python3 ./scripts/vv8cli.py docker -f` do you see any errors?

**Comment by aswad546** [comment]
I chose local, default 80 instances and built the post processors during setup. I am attaching the output of the command you sent.

[log.txt](https://github.com/user-attachments/files/15779985/log.txt)

**Comment by sohomdatta1** [comment]
It seems like whatever CPU architecture you are on doesn't support running MongoDB containers since the MongoDB container exits with:
```
mongodb-1            | /usr/local/bin/docker-entrypoint.sh: line 416:    26 Illegal instruction     (core dumped) "${mongodHackedArgs[@]}" --fork
mongodb-1            | 
mongodb-1            | WARNING: MongoDB 5.0+ requires a CPU with AVX support, and your current system does not appear to have that!
mongodb-1            |   see https://jira.mongodb.org/browse/SERVER-54407
mongodb-1            |   see also https://www.mongodb.com/community/forums/t/mongodb-5-0-cpu-intel-g4650-compatibility/116610/2
mongodb-1            |   see also https://github.com/docker-library/mongo/issues/485#issuecomment-891991814
```
You could try pinning the MongoDB version to a older release by editing https://github.com/wspr-ncsu/visiblev8-crawler/blob/9f65b3bc134a159bc2f3664ad1ee82a985ce980c/docker-compose.yaml#L55. [StackOverflow seems to think](https://stackoverflow.com/questions/70818543/mongo-db-deployment-not-working-in-kubernetes-because-processor-doesnt-have-avx) that `mongo:4.4.18` doesn't require AVX.

**Comment by aswad546** [comment]
That seemed to fix the issue, thanks!

**Comment by Ale0x78** [comment]
Also! If you are running this inside a virtual machine, you might want to check to make sure the CPU type is set to something that supports AVX 

[How to enable AVX support in VCenter](https://community.broadcom.com/vmware-cloud-foundation/discussion/how-to-enableuse-avxavx2-cpu-instructions-in-esxi-environment-there-isnt-vcenter)
[How to enable AVX support in proxmox](https://forum.proxmox.com/threads/enable-avx.129019/)

**Comment by aswad546** [comment]
Thanks so much, it seems the machine I am currently using does not have hardware support for AVX instructions.

**Comment by aswad546** [comment]
I did have a question, given I am using a VM inside a remote machine is there some way for me to set headless mode of the puppeteer crawler to false and view and interact with the GUI (within your system)

**Comment by aswad546** [comment]
Hello Sohom,

Where do you route the error or debug statements that the crawler workers collect?

Best,
Aswad

**Comment by sohomdatta1** [comment]
> Hello Sohom,
> 
> Where do you route the error or debug statements that the crawler workers collect?
> 
> Best, Aswad

What do you mean by debug/error statements? The number of crawls that errored out can be viewed on the flower UI and/or by using the `/status/` API through the fetch command.

**Comment by aswad546** [comment]
I want to see the reason for why particular crawls errored out. Do you store the error trace somewhere?

**Comment by sohomdatta1** [comment]
Sorry for the late reply, we don't store the error trace anywhere right now since based on our experiences the failures tend to be timeout/connection errors from due to the fact that the website doesn't exist. If you are doing smaller crawls, you can view the error traces in the docker logs.

**Comment by sohomdatta1** [comment]
If there is a usecase of keeping the error logs around, we could add it to the crawler (it's a few lines of python code in https://github.com/wspr-ncsu/visiblev8-crawler/blob/main/celery_workers/vv8_worker/tasks.py#L59 that need to be changed) :)

**Comment by aswad546** [comment]
Thanks for this, I was also wondering while designing the crawler did you all consider evading bot detection scripts?

**Comment by sohomdatta1** [comment]
> Thanks for this, I was also wondering while designing the crawler did you all consider evading bot detection scripts?

We did consider bot detection scripts, in the crawler we are using `puppeteer-extra-stealth-plugin` to provide some level of evasion from detection scripts. It isn't perfect, (Cloudflare in particular is not bypassed) but for most normal crawling (`tranco-10k`) we've found it to be sufficient.

**Comment by aswad546** [comment]
Thank you!

**Comment by aswad546** [comment]
Hey, I wanted to ask if the crawler supports running the workers in headful mode using Xvfb? If so how can I trigger this? I noticed the vv8_worker.dockerfile has mentions of using vnc and xvfb.

**Comment by sohomdatta1** [comment]
You should be able to just pass the `--no-headless` parameter to the existing crawl command to use the headful mode. We don't use xvfb since we have had issues with how Chrome reacts to Xvfb viewport sizes (and the subsequent fingerprintability of that interaction), instead, we include a minimal XFCE environment that can be viewed using VNC at port 0.0.0.0:5901 and 0.0.0.0:6901.

**Comment by aswad546** [comment]
Thank you that seems to work well!

**Comment by aswad546** [comment]
Hello again,

I am trying to crawl multiple websites using this crawler it seems a lot of websites are facing navigation timeouts and the cpu seems to be overloaded. I am using 32 cores for now and 256GB of RAM for Ubuntu 22.04. I have set the number of parallel instances to be 32 instead of the suggested 128. And my navigation timeout is quite high. If I run these websites individually they are successfully crawled but in this case it seems they are failing. I did examine the network load and it seems that is not the issue. Is this usual, any suggestions? It will be hard to scale this way. Any recommendations?

Thanks,
Aswad
